### DISCUSSION TITLE
Question: When does `ConcurrentDictionary<TKey, TValue>.TryRemove(TKey, out TValue)`

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/runtime#97994] [General] [answered]
I'm trying to use a concurrent dictionary for thread-safe access, and although it's not the main concern, high performance and low code complexity are something I'm interested in.

Looking at the documentation for [ConcurrentDictionary<TKey, TValue>.TryRemove(TKey, out TValue)](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryremove?view=net-8.0#system-collections-concurrent-concurrentdictionary-2-tryremove(-0-1@)), The comment about the return value is "`true` if the object was removed successfully; otherwise, `false`".

What does "removed successfully" mean? Does it mean the key was found, or is it possible for a key to exist while being failed to be removed? It's not inherently clear to me from this phrasing.

Right now I'm being defensive and writing code like this:

```csharp
internal void Clear(string context)
{
    while (_outputs.ContainsKey(context))
    {
        if (_outputs.TryRemove(context, out StringWriter? writer))
        {
            writer.Dispose();
        }
    }
}
```

And I want to know if it's still safe to write

```csharp
internal void Clear(string context)
{
    if (_outputs.TryRemove(context, out StringWriter? writer))
    {
        writer.Dispose();
    }
}
```

If possible, could the documentation be changed so its more clear to anyone not familiar enough with `ConcurrentDictionary<TKey, TValue>` internals to know this?

**Comment by ghost** [comment]
Tagging subscribers to this area: @dotnet/area-system-collections
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.
<details>
<summary>Issue Details</summary>
<hr />

I'm trying to use a concurrent dictionary for thread-safe access, and although it's not the main concern, high performance and low code complexity are something I'm interested in.

Looking at the documentation for [ConcurrentDictionary<TKey, TValue>.TryRemove(TKey, out TValue)](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryremove?view=net-8.0#system-collections-concurrent-concurrentdictionary-2-tryremove(-0-1@)), The comment about the return value is "`true` if the object was removed successfully; otherwise, `false`".

What does "removed successfully" mean? Does it mean the key was found, or is it possible for a key to exist while being failed to be removed? It's not inherently clear to me from this phrasing.

Right now I'm being defensive and writing code like this:

```csharp
internal void Clear(string context)
{
    while (_outputs.ContainsKey(context))
    {
        if (_outputs.TryRemove(context, out StringWriter? writer))
        {
            writer.Dispose();
        }
    }
}
```

And I want to know if it's still safe to write

```csharp
internal void Clear(string context)
{
    if (_outputs.TryRemove(context, out StringWriter? writer))
    {
        writer.Dispose();
    }
}
```

If possible, could the documentation be changed so its more clear to anyone not familiar enough with `ConcurrentDictionary<TKey, TValue>` internals to know this?

<table>
  <tr>
    <th align="left">Author:</th>
    <td>RenderMichael</td>
  </tr>
  <tr>
    <th align="left">Assignees:</th>
    <td>-</td>
  </tr>
  <tr>
    <th align="left">Labels:</th>
    <td>

`area-System.Collections`, `untriaged`

</td>
  </tr>
  <tr>
    <th align="left">Milestone:</th>
    <td>-</td>
  </tr>
</table>
</details>

**Reply by stephentoub** [reply]
If it's found, it'll be removed.  (Of course, concurrently someone else could add it back.)

**Reply by RenderMichael** [reply]
What if it exists but it's not found? The documentation makes it not perfectly clear for me.

Basically what I'm asking is whether the second code block is just as correct as the first code block.

**Reply by stephentoub** [reply, marked as answer]
If it exists, it'll be found and removed and true will be returned.
